synedrion/src/cggmp21/key_refresh.rs

```diff
@@ -1,321 +1,632 @@
+                verify_that(!r3_bc.psi_prime.verify(&paillier_pk, &aux))
+            }
+            Error::R3FacFailed { reported_by } => {
+                let rid = reconstruct_rid::<P, _>(deserializer, &previous_messages, &combined_echos)?;
+                let aux = (&sid, guilty_party, &rid);
+
+                let r2_eb = combined_echos
+                    .get_round(2)?
+                    .try_get("combined echos for Round 2", reported_by)?
+                    .deserialize::<Round2EchoBroadcast<P, Id>>(deserializer)?;
+                let r2_bc = previous_messages
+                    .get_round(2)?
+                    .normal_broadcast
+                    .deserialize::<Round2NormalBroadcast<P, Id>>(deserializer)?;
+                let r3_dm = message
+                    .direct_message
+                    .deserialize::<Round3DirectMessage<P>>(deserializer)?;
+                let paillier_pk = r2_bc.paillier_pk.into_precomputed();
+                let rp_params = r2_eb.rp_params.to_precomputed();
+                verify_that(!r3_dm.psi.verify(&paillier_pk, &rp_params, &aux))
+            }
+            Error::R3WrongIdsHatPsi => {
+                let r3_eb = message
+                    .echo_broadcast
+                    .deserialize::<Round3EchoBroadcast<Id>>(deserializer)?;
+                verify_that(r3_eb.hat_psis.keys().cloned().collect::<BTreeSet<_>>() != associated_data.ids)
+            }
+            Error::R3SchFailed { failed_for } => {
+                let rid = reconstruct_rid::<P, _>(deserializer, &previous_messages, &combined_echos)?;
+                let aux = (&sid, guilty_party, &rid);
+
+                let r2_bc = previous_messages
+                    .get_round(2)?
+                    .normal_broadcast
+                    .deserialize::<Round2NormalBroadcast<P, Id>>(deserializer)?;
+                let r3_eb = message
+                    .echo_broadcast
+                    .deserialize::<Round3EchoBroadcast<Id>>(deserializer)?;
+
+                let cap_a = r2_bc.cap_as.try_get("Schnorr commitments", failed_for)?;
+                let cap_x = r2_bc.cap_xs.try_get("public share changes", failed_for)?;
+                let hat_psi = r3_eb.hat_psis.try_get("Schnorr proofs", failed_for)?;
+                verify_that(!hat_psi.verify(cap_a, cap_x, &aux))
+            }
+        }
+    }
+}
+
+#[derive(Debug, Clone)]
+pub(super) struct PublicData<P: SchemeParams, Id> {
+    pub(super) cap_xs: BTreeMap<Id, Point>, // $X_{i,j}$ where $i$ is this party's index
+    pub(super) cap_ys: BTreeMap<Id, Point>, // $Y_{i,j}$ where $i$ is this party's index
+    pub(super) cap_as: BTreeMap<Id, SchCommitment>, // $A_{i,j}$ where $i$ is this party's index
+    pub(super) paillier_pk: PublicKeyPaillier<P::Paillier>, // $N_i$
+    pub(super) rp_params: RPParams<P::Paillier>, // $\hat{N}_i$, $s_i$, and $t_i$
+    pub(super) psi: PrmProof<P>,
+    rid: BitVec,
```

Comment:
DQ: is there a more descriptive name for `rid` we can use? It's not meant to be read "round id" is it?

Reply:
That's what the paper uses, so in the spirit of preserving variable names, this one is preserved as well. I think it's supposed to be "refresh ID"